Add plugin: Boardgame Search

Marlon154 opened the pull request:
add obsidian-boardgame-plugin

# I am submitting a new Community Plugin

## Repo URL

<!--- Paste a link to your repo here for easy access -->
Link to my plugin: https://github.com/Marlon154/boardgame-search

## Release Checklist
- [x] I have tested the plugin on
  - [ ]  Windows
  - [ ]  macOS
  - [x]  Linux
  - [x]  Android _(if applicable)_
  - [x]  iOS _(if applicable)_
- [x] My GitHub release contains all required files (as individual files, not just in the source.zip / source.tar.gz)
  - [x] `main.js`
  - [x] `manifest.json`
  - [x] `styles.css` _(optional)_
- [x] GitHub release name matches the exact version number specified in my manifest.json (_**Note:** Use the exact version number, don't include a prefix `v`_)
- [x] The `id` in my `manifest.json` matches the `id` in the `community-plugins.json` file.
- [x] My README.md describes the plugin's purpose and provides clear usage instructions.
- [x] I have read the developer policies at https://docs.obsidian.md/Developer+policies, and have assessed my plugins's adherence to these policies.
- [x] I have read the tips in https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines and have self-reviewed my plugin to avoid these common pitfalls.
- [x] I have added a license in the LICENSE file.
- [x] My project respects and is compatible with the original license of any code from other plugins that I'm using.
      I have given proper attribution to these other projects in my `README.md`.

Zachatoo commented:
[# Obsidian Board Game Tracker](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/README.md#L3)
Naming something "Obsidian xyz" is reserved for first party products we create, please change this header.

["description": "An Obsidian plugin that helps board game enthusiasts track their game collection seamlessly within their notes.",](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/manifest.json#L6)
Avoid including sentences like "This is a plugin for Obsidian that does ..." in your description, it should be a given since this is only shown when browsing plugins through the app or on our website.
We have published a guide for plugin descriptions to follow here: https://docs.obsidian.md/Plugins/Releasing/Submission+requirements+for+plugins#Keep+plugin+descriptions+short+and+simple.

[await this.app.vault.adapter.mkdir(this.settings.imagePath);](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/main.ts#L88-L91), [await this.app.vault.adapter.mkdir(this.settings.folder);](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/main.ts#L147)
[Prefer the Vault API over the Adapter API](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Prefer+the+Vault+API+over+the+Adapter+API)

[const filePath = `${this.settings.imagePath}/${fileName}`;](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/main.ts#L85)
[Use `normalizePath` to clean up user defined paths](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+%60normalizePath()%60+to+clean+up+user-defined+paths)

[const response = await fetch(`${this.baseUrl}/search?query=${encodeURIComponent(query)}&type=boardgame`);](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/api/BGGApiManager.ts#L66), [const response = await fetch(`${this.baseUrl}/thing?id=${gameId}&stats=1`);](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/api/BGGApiManager.ts#L206)
Use the `requestUrl` function from the Obsidian API instead, it will handle some things like CORS automatically.

[containerEl.createEl('h2', { text: 'File Settings' });](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L22), [containerEl.createEl('h2', { text: 'Image Settings' });](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L64), [containerEl.createEl('h2', { text: 'Session Tracking' });](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L101), [containerEl.createEl('h2', { text: 'BoardGameGeek Integration' });](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L119), [containerEl.createEl('h2', { text: 'General Settings' });](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L144), [containerEl.createEl('h2', { text: 'Chart Settings' });](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L167)
[Avoid "settings" in settings headings](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+%22settings%22+in+settings+headings)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)
For section headings in settings use:
```ts
new Setting(containerEl).setName('name here').setHeading();
```

[.setName('Show Images in Search')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L67), [.setName('Save Images Locally')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L77), [.setName('Image Save Location')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L87), [.setName('Enable Session Tracking')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L106), [.setName('Enable BGG Integration')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L122), [.setName('BGG Username')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L132), [.setName('Open Page After Creation')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L147), [.setName('Use Charts Plugin')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L170), [.setName('Chart Width')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L180), [.setName('Track Players')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L201), [.setName('Track Scores')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L211), [.setName('Track Play Time')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L221), [.setName('Track Winners')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L231)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)

[containerEl.createEl('h2', { text: 'General Settings' });](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L144)
[Don't add a top-level heading in the settings tab, such as "General", "Settings", or the name of your plugin.](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Only+use+headings+under+settings+if+you+have+more+than+one+section.) These settings should not be under a heading and should be before all other settings. I would consider putting the settings under the "file settings" heading as general settings and not being under a heading as well, I'll leave that up to you to decide.

Whenever you change a setting that is affected by disabled logic of other settings (for example, when changing the "save images locally" setting), make sure you call `this.display()` to re-render the settings tab so fields are properly enabled disabled. I also recommend hiding settings instead of disabling them to avoid confusion when a user tries to update them, the disabled state for some settings don't look very different from the enabled state.

[export abstract class TextInputSuggest\<T> implements ISuggestOwner\<T> {](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/suggesters/suggest.ts#L97)
There's a [AbstractInputSuggest](https://docs.obsidian.md/Reference/TypeScript+API/AbstractInputSuggest) provided to you in the Obsidian API, I don't believe you need to implement this yourself. The example you're pulling this from is outdated. You should be able to remove this file and your dependency on `@popperjs/core`. Please correct me if I'm wrong.

[const inputBeforeCursor = inputStr.substr(cursorPosition - lookbehind, lookbehind);](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/suggesters/FileNameFormatSuggester.ts#L30), [return \`${currentInputValue.substr(](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/suggesters/FileNameFormatSuggester.ts#L48-L51)
[`.substr`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substr) is deprecated, please use [`.substring`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/String/substring#the_difference_between_substring_and_substr) instead.

[console.log('No template path provided, using default template');](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/template/TemplateManager.ts#L74)
Please avoid unnecessary logging. If you need it for debugging purposes, add a check to only log during development. Your other logs seem fine.

[let debounceTimeout: NodeJS.Timeout;](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/main.ts#L201), [debounceTimeout = setTimeout(async () => {](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/main.ts#L212), [clearTimeout(debounceTimeout);](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/main.ts#L203)
I'm not sure if `NodeJS.Timeout` is available on mobile, I recommend using `window.setTimeout` and `window.clearTimeout` instead.

Make sure you are following the [BGG XML API terms of use](https://boardgamegeek.com/wiki/page/XML_API_Terms_of_Use), specifically this section where they want you to include their logo. I suspect adding the logo in your README will suffice.

---

[const abstractFiles = this.app.vault.getAllLoadedFiles();](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/suggesters/FileSuggester.ts#L8)
Since you only care about markdown files here, you can use [`getMarkdownFiles`](https://docs.obsidian.md/Reference/TypeScript+API/Vault/getMarkdownFiles) instead to get all markdown files in the vault.

[cb.setPlaceholder('Example: Games')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L29), [cb.setPlaceholder('Example: templates/game')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L55), [cb.setPlaceholder('Example: Games')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L91), [cb.setPlaceholder('Example: {{gameName}}')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L42)
No need to prefix placeholder text with "Example :". It feels out of place with other Obsidian UI.

[.setName('Overwrite exsisting files')](https://github.com/Marlon154/boardgame-search/blob/8f3ca29e95616e1a346445ede1dcf72ccc2062a4/src/settings/SettingTab.ts#L157)
Minor typo, should be "existing".

In the default template, the play time and the year published are on the same line, probably has something to do with how nunjucks is handling trailing line breaks after a `{% endif %}`. There also appears to be some missing data sanitization for mdash. See example below for the Harmonies board game.
```
## Game Details
- **Min Players:** 1
- **Max Players:** 4
- **Play Time:** 45 minutes- **Year Published:** 2024
- **BGG Rating:** 8.0/10
## Description
In Harmonies, build landscapes by placing colored tokens and create habitats for your animals. To earn the most points and win the game, incorporate the habitats in your landscapes wisely and have as many animals as you can settle there.

&mdash;description from the publisher
```

---

As an avid board gamer myself, this is awesome, thanks for building this!

Marlon154 commented:
Thanks for the thorough review! Really appreciate the attention to detail, particularly with the text sanitization issues in the template, and the heads up about BGG's API terms of use. I've made the following changes:

- Added BGG attribution and linked logo to comply with API terms
- Extended sanitization
- Fixed default template line breaks
- Simplified suggesters by using Obsidian's `AbstractInputSuggest`
- Fixed all vault operations to use proper Vault API instead of adapter
- Added proper type for timeouts using `window.setTimeout`
- Updated settings UI to follow Obsidian guidelines

Let me know if you spot anything else that needs addressing!

Zachatoo commented:
[const folder = this.app.vault.getAbstractFileByPath(folderPath);](https://github.com/Marlon154/boardgame-search/blob/2c077460d6a8312d8476b48ddf7bf05887e8367d/src/main.ts#L88), [const file = this.app.vault.getAbstractFileByPath(filePath);](https://github.com/Marlon154/boardgame-search/blob/2c077460d6a8312d8476b48ddf7bf05887e8367d/src/main.ts#L93), [const existingFile = this.app.vault.getAbstractFileByPath(filePath);](https://github.com/Marlon154/boardgame-search/blob/2c077460d6a8312d8476b48ddf7bf05887e8367d/src/main.ts#L139), [const folder = this.app.vault.getAbstractFileByPath(folderPath);](https://github.com/Marlon154/boardgame-search/blob/2c077460d6a8312d8476b48ddf7bf05887e8367d/src/main.ts#L152)
You can use [getFileByPath](https://docs.obsidian.md/Reference/TypeScript+API/Vault/getFileByPath) and [getFolderByPath](https://docs.obsidian.md/Reference/TypeScript+API/Vault/getFolderByPath), depending on the situation, since `getAbstractFileByPath` can return a `TFile` or a `TFolder`. In some cases you are accounting for whether it's a file or folder or null, in others you are not. I'd recommend moving over to these methods so you can just do a null check.

[.setName('Use Charts Plugin')](https://github.com/Marlon154/boardgame-search/blob/2c077460d6a8312d8476b48ddf7bf05887e8367d/src/settings/SettingTab.ts#L171), [.setName('Use Charts Plugin')](https://github.com/Marlon154/boardgame-search/blob/2c077460d6a8312d8476b48ddf7bf05887e8367d/src/settings/SettingTab.ts#L171)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI) (Charts can be uppercase, Plugin should not be)

[new Setting(containerEl).setName('Chart settings').setHeading();](https://github.com/Marlon154/boardgame-search/blob/2c077460d6a8312d8476b48ddf7bf05887e8367d/src/settings/SettingTab.ts#L168)
[Avoid "settings" in settings headings](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Avoid+%22settings%22+in+settings+headings)

["minAppVersion": "0.15.0",](https://github.com/Marlon154/boardgame-search/blob/2c077460d6a8312d8476b48ddf7bf05887e8367d/manifest.json#L5)
You are using newer API's, set this to the latest public build number.

---

Looks like the formatting of the tables got broken, I think you may have removed too many line breaks.

Marlon154 commented:
Thanks for the review! I'll fix:

- Update file/folder methods to use getFileByPath/getFolderByPath
- Fix settings text
- Update minAppVersion
- Fix default template formatting

Let me know if you'd like to see any specific changes.

joethei commented:
[class BGGSearchModal extends Modal {](https://github.com/Marlon154/boardgame-search/blob/dd5d73a0a6c35f8ff1a51bba94c66d792e9ba8e6/src/main.ts#L170)
Instead of a regular modal this should be either a `SuggestModal` or a `FuzzySuggestModal` 
See some example code here: https://docs.obsidian.md/Plugins/User+interface/Modals#Select+from+list+of+suggestions

[export const BGGLOGOIMGAEURL = '](https://github.com/Marlon154/boardgame-search/blob/dd5d73a0a6c35f8ff1a51bba94c66d792e9ba8e6/src/constants.ts#L6)
Static resources like that should be bundled inside the plugin and not loaded from an external source.
Some docs: https://docs.obsidian.md/Themes/App+themes/Embed+fonts+and+images+in+your+theme

['Note Creation'](https://github.com/Marlon154/boardgame-search/blob/dd5d73a0a6c35f8ff1a51bba94c66d792e9ba8e6/src/settings/SettingTab.ts#L79)
[Use sentence case in UI](https://docs.obsidian.md/Plugins/Releasing/Plugin+guidelines#Use+sentence+case+in+UI)
